src/adapter/src/coord/read_policy.rs

```diff
@@ -713,67 +717,30 @@ impl crate::coord::Coordinator {
         &mut self,
         id_bundle: &CollectionIdBundle,
     ) -> ReadHolds<Timestamp> {
-        // Create a `ReadHolds` that contains a read hold for each id in
-        // `id_bundle`. The time of each read holds is at `time`, if possible
-        // otherwise it is at the lowest possible time.
-        //
-        // This does not apply the read holds in COMPUTE. The code below applies
-        // those in the correct read capability.
         let mut read_holds = ReadHolds::new();
-        let time_antichain = Antichain::from_elem(Timestamp::MIN);
 
         let desired_storage_holds = id_bundle.storage_ids.iter().map(|id| *id).collect_vec();
         let storage_read_holds = self
             .controller
             .storage
             .acquire_read_holds(desired_storage_holds)
-            .expect("missing collections");
-
-        for storage_read_hold in storage_read_holds {
-            let prev = read_holds
-                .storage_holds
-                .insert(storage_read_hold.id(), storage_read_hold);
-
-            assert!(
-                prev.is_none(),
-                "can only store one storage ReadHold per collection"
-            );
-        }
+            .expect("missing storage collections");
+        read_holds.storage_holds = storage_read_holds
+            .into_iter()
+            .map(|hold| (hold.id(), hold))
+            .collect();
 
-        for (compute_instance, compute_ids) in id_bundle.compute_ids.iter() {
-            for id in compute_ids.iter() {
-                let collection = self
+        for (&instance_id, collection_ids) in &id_bundle.compute_ids {
+            for &id in collection_ids {
+                let hold = self
                     .controller
                     .compute
-                    .collection(*compute_instance, *id)
-                    .expect("collection does not exist");
-                let read_frontier = collection.read_capability().clone();
-                let time_antichain = time_antichain.join(&read_frontier);
-                let hold_chain = MutableAntichain::from(time_antichain);
-                read_holds
-                    .compute_holds
-                    .insert((*compute_instance, *id), hold_chain);
+                    .acquire_read_hold(instance_id, id)
+                    .expect("missing compute collection");
+                read_holds.compute_holds.insert((instance_id, id), hold);
```

Comment:
The assumption is that each compute collection is installed on a single cluster, and that collection IDs are globally unique, so a compute ID cannot occur more than once in a `CollectionIdBundle`. You can of course build a `CollectionIdBundle` where this is violated, so it's reasonable to have an assert here.

Reply:
I'm more worried about someone calling this method multiple times with the same ID, rather than a single time with a malformed `CollectionIdBundle`.

For example `initialize_read_policies` makes it clear IMO that it should only be called once, after a collection is created. 

>     /// Initialize the storage and compute read policies.
>     ///
>     /// This should be called only after a collection is created, and
>     /// ideally very soon afterwards.

However from looking at `acquire_read_holds` it's not clear that it's only expected to be called a single time per collection. Though when looking at the implementation it seems like that's the intention.

To be honest, I'm having trouble understanding the difference between `initialize_read_policies` and `acquire_read_holds`, and when someone should call one over the other.

Reply:
Ah, OK I see. I'm just completely misreading this method. `read_holds` is not a field on `self` but a new object. Never mind then.

Reply:
Yep! `acquire_read_holds` just collects the specified read holds from the controllers. `initialize_read_policies` calls it to collect read holds, then installs them in the timeline state, then installs read policies.

Reply:
Added an assert for unique compute collection IDs.